test/controllers/articles_controller_test.rb

```diff
@@ -3,85 +3,143 @@
 require "test_helper"
 
 # Controller tests for news articles
-class ArticleControllerTest < FunctionalTestCase
-  def test_create_article_get
+class ArticlesControllerTest < FunctionalTestCase
+  ############ test Actions that Display data (index, show, etc.)
+
+  def test_index
+    get(:index)
+    assert(:success, "Any user should be able to see article index")
+    Article.find_each do |article|
+      assert_select(
+        "a[href *= '#{article_path(article.id)}']", true,
+        "full Article Index missing link to #{article.title} (##{article.id})"
+      )
+    end
+
+    article = Article.first
+    query = Query.lookup(:Article, :in_set, ids: [article.id])
+    params = @controller.query_params(query)
+    get(:index, params: params)
+    assert_select("a:match('href',?)", %r{/articles/\d+}, { count: 1 },
+                  "filtered Article Index has wrong number of entries")
+    assert_select(
+      "a[href *= '#{article_path(article.id)}']", true,
+      "filtered Article Index missing link to #{article.title} (##{article.id})"
+    )
```

Comment:
Are these `assert_select`s un-performant? If so, what should I do?

Reply:
I'd time it with the assert_selects and without.  If the different is less than a second or less than 10% of the run time for the rest of the test I wouldn't sweat it.  I'm guessing the `get` dominates the time it takes to run the test.